docs/man/man8/echkpnt.8
.ds ]W %
.ds ]L
.nh
.TH echkpnt  8 "November 2003 - Platform LSF"
.br

.SH NAME 
\fBechkpnt \fR  - checkpoints a process or process group of an LSF batch job 
.SH SYNOPSIS 
.BR
.PP
.PP
\fBechkpnt\fR [\fB-c\fR] [\fB-f\fR] [\fB-k\fR | \fB-s\fR] [\fB-d\fR \fIcheckpoint_dir\fR] [\fB-x\fR] 
\fIprocess_group_ID\fR 
.PP
\fBechkpnt\fR [\fB-h\fR | \fB-V\fR] 
.SH DESCRIPTION 
.BR
.PP
.PP
echkpnt is the interface through which jobs are checkpointed. 
.PP
By default, jobs continue to execute after they have been checkpointed. 
.PP
Some options in the synopsis are only supported by some operating systems. 
.SS echkpnt 
.BR
.PP

.IP
echkpnt is located in LSF_SERVERDIR. 

.IP
echkpnt sends checkpoint instructions provided through job submission 
or through queue parameters to LSF's checkpoint facility 
echkpnt.default or to any custom echkpnt methods that you define. 
Communication is achieved through the common syntax indicated in the 
synopsis. 


.SS echkpnt.default 
.BR
.PP

.IP
echkpnt.default is located in LSF_SERVERDIR. 

.IP
By default, when a job is checkpointed, echkpnt.default is used. 
echkpnt.default uses the arguments \fIcheckpoint_dir\fR and 
process_group_ID to identify where checkpoint files are to be created 
and which job is to be checkpointed. echkpnt.default then creates 
checkpoint files in\fI \fR\fIcheckpoint_dir\fR/\fIjob_ID\fR and attempts to checkpoint 
the job. 


.SS echkpnt.method_name 
.BR
.PP

.IP
echkpnt.\fImethod_name\fR is a custom echkpnt that attempts to run a custom 
checkpoint program or script. 

.IP
The name of the custom echkpnt method must be 
echkpnt.\fImethod_name\fR, where \fImethod_name\fR is a name you assign to 
your custom echkpnt. The method_name must be specified either with 
LSB_ECHKPNT_METHOD (in lsf.conf, as an environment variable) 
and/or when the job is submitted with the bsub -k option such as bsub -
k "mydir 120 method=myapp" job1. 

.IP
echkpnt.\fImethod_name\fR must support the syntax described in the 
synopsis. 

.IP
echkpnt.\fImethod_name\fR is placed in LSF_SERVERDIR, or in the directory 
specified by LSB_ECHKPNT_METHOD_DIR (set in lsf.conf or as an 
environment variable). 

.IP
The method name (LSB_ECHKPNT_METHOD in lsf.conf or as an 
environment variable) and location (LSB_ECHKPNT_METHOD_DIR) 
combination must be unique in the cluster. For example, you may have 
two echkpnt applications with the same name such as 
echkpnt.mymethod but what differentiates them is the different directories 
defined with LSB_ECHKPNT_METHOD_DIR. 

.IP
The checkpoint method directory should be accessible by all users who 
need to run the custom echkpnt and erestart programs.

.IP
Do not use default as a method name as this is a reserved keyword to 
indicate to use LSF's default echkpnt and erestart methods. 

.IP
If you want to specify to use LSF's echkpnt.default for a specific 
application, specify method=default at job submission with bsub -k. 


.SH OPTIONS 
.BR
.PP
.TP 
\fB-c\fR 

.IP
This option is supported on only some operating systems such as Convex. 

.IP
Copies to the checkpoint directory all files in use by the checkpointed process. 


.TP 
\fB-f\fR 

.IP
This option is supported on only some operating systems such as Convex. 

.IP
Forces a job to be checkpointed even if non-checkpointable conditions exist 
(these conditions are specific to the type of checkpoint facility being used). 
This option may create checkpoint files that will not restart properly. 


.TP 
\fB-k\fR | \fB-s\fR 

.IP
Options -k and -s are mutually exclusive. 


.RE
.IP
\fB-k\fR 
.BR
.RS

.IP
Kills a job after it has been successfully checkpointed. If this option is 
specified and the checkpoint fails for any reason, the job continues to 
execute. 



.RE
.IP
\fB-s\fR 
.BR
.RS

.IP
This option is supported on only some operating systems such as SGI 
systems running IRIX, and Convex. 

.IP
Stops a job after it has been successfully checkpointed. If this option is 
specified and the checkpoint fails for any reason, the job continues to 
execute. 



.TP 
\fB-d\fR \fIcheckpoint_dir\fR 

.IP
Specifies the checkpoint directory. Specify a relative or absolute path name. 

.IP
When a job is checkpointed, the checkpoint information is stored in 
\fIcheckpoint_dir\fR/\fIjob_ID\fR/\fIfile_name\fR. Multiple jobs can checkpoint into the 
same directory. The system can create multiple files. 

.IP
The checkpoint directory is used for restarting the job (see brestart(1)).


.TP 
\fB-x\fR 

.IP
This option is supported on only some operating systems such as SGI systems 
running IRIX. 

.IP
Identifies the cpr process as type HID (process hierarchy rooted at that PID). 


.TP 
\fIprocess_group_ID\fR 

.IP
ID of the process or process group to be checkpointed. 


.TP 
\fB-h\fR 

.IP
Prints command usage to stderr and exits. 


.TP 
\fB-V\fR 

.IP
Prints LSF release version to stderr and exits. 


.SH SEE ALSO 
.BR
.PP
.PP
bsub(1), bchkpnt(1), brestart(1), erestart(8) 
.SH DIAGNOSTICS 
.BR
.PP
.SS echkpnt 
.BR
.PP

.IP
Exits with a 0 if the checkpoint operation succeeds. Otherwise, exits with 
the value of echkpnt.default or echkpnt.\fImethod_name\fR. 


.SS echkpnt.default 
.BR
.PP

.IP
If LSF's default checkpoint program echkpnt.default is used and the job 
is successfully checkpointed, echkpnt.default exits with a 0 and writes 
nothing to stdout and stderr or the message "Done" or "Success". 
echkpnt.default returns non-zero values if the checkpoint operation 
fails. In addition, all messages in stdout or stderr that are returned by 
echkpnt.default are interpreted by LSF as failure of the checkpoint 
operation. 


.SS echkpnt.method_name 
.BR
.PP

.IP
If a custom checkpoint program is used (LSB_ECHKPNT_METHOD is 
defined in lsf.conf or as an environment variable, or the checkpoint 
method specified such as bsub -k "mydir method=myapp" job1), 
echkpnt.\fImethod_name\fR exits with a 0 if it succeeds in checkpointing the 
job. Non-zero values indicate job checkpoint failed. All messages written 
to stdout and stderr are directed to /dev/null and ignored by LSF. 

.IP
To save standard error and standard out messages for echkpnt. 
\fImethod_name\fR, set LSB_ECHKPNT_KEEP_OUTPUT=y in lsf.conf or as 
an environment variable. The stdout and stderr output generated by 
echkpnt. \fImethod_name\fR will be redirected to: 

.IP
- \fIcheckpoint_dir\fR/$LSB_JOBID/echkpnt.out 

.IP
- \fIcheckpoint_dir\fR/$LSB_JOBID/echkpnt.err 


.SH LIMITATIONS 
.BR
.PP
.PP
If you define LSF_ECHKPNTDIR, you cannot use a custom echkpnt method. It 
is recommended to no longer use LSF_ECHKPNTDIR but use instead 
LSB_ECHKPNT_METHOD and LSB_ECHKPNT_METHOD_DIR. 
.PP
If you use echkpnt.\fImethod_name\fR, once the job has been submitted, the job 
will be checkpointed with the method that was specified at job submission or 
with the parameter LSB_ECHKPNT_METHOD. You cannot change the method 
with the bmod command. 
.PP
If you submit a job and do not specify a custom method, and 
LSB_ECHKPNT_METHOD is not defined, echkpnt.default will be used. You 
will not be able to change this with bmod. 
.PP
It is the cluster administrator's responsibility to ensure that method name and 
method directory combinations are unique in the cluster.

.\" Generated by Quadralay WebWorks Publisher Professional Edition 6.0.7
.\" Based on template Portable HTML
.\" Copyright 1994-2003 Platform Computing Corporation


ChangeLog
2012-01-23  David Bigagli  <david@openlava.net>

	* Created 2.0 branch.

2011-11-07  David Bigagli  <david@openlava.net>

	* lib.host.c Rewrote the caching code to read the hosts file only
	once and add the host in the cache. Removed the cache timeout
	and simplified the entire implementation to use only two interface
	functions: Gethostbyname_() and Gethostbyaddr_().
	Change the old C types u_int and casta to char *
	into the current linux types whenever required.
	Modified the daemons code to use the new host library
	implementation.

2011-10-30  David Bigagli  <david@openlava.net>

	* mbd.policy.c Developed a little round robin scheduler absed on the number of jobs a user has running.

	* openlava.spec  modified the spec file so that autoconf creates
	openlava configuration files and use the outptu variables to make
	the necessary subsititution in the them. Change the post install
	to just erase the package without saving anything.
	Removed the symbolic link as that is something sites have to
	do as they may want to run more versions together, also
	in now the lsf.conf has the version in the openlava
	fundamental variables clearly indicating which version is in use.

2011-09-16  David Bigagli  <david@openlava.net>

	* README removed obsolete information about the source code directory structure.
	Added the README and COPYING file to the rmp file.

2011-09-14  David Bigagli  <david@openlava.net>

	* lsf.h Added the openlava foundation copyright.

2011-09-09  David Bigagli  <david@openlava.net>

	* openlava.spec Updated the spec file to reflect the changes in teh build system. The ownership of all files changed, no longer root, the entire installation tree is owned by the openlava user. The daemons to run in multi user mode must still be started at root which is guaranteed at boot by init.

2011-08-27  David Bigagli  <david@openlava.net>

	* openlavaconfig.h Removed it from the top directory and merge its content into lsf.h which is the openlava master header file. In this way the repository is more GNU like as well.

2011-08-23  David Bigagli  <david@openlava.net>

	* GNU build system. Switched openlava from home made makefiles to the GNU build system.

2011-07-20  David Bigagli  <david@openlava.net>

	* scripts/Makefile: added the installation
	of scripts that manage the integration with
	mpi and openmpi.

2011-07-17  David Bigagli  <david@openlava.net>

	* lib.table.c: Added getClosestPrime() function to
	always size and resize tables as primies.

	* lsbatch/daemons/mbd.init.c (addHost):
	Set the number of CPUs on batch host to be its MXJ.
	If MXJ is not confgiured than use the number of physical
	CPUs as provided by openlava base.


2011-07-16  David Bigagli  <david@openlava.net>

	* Added lsb.hosts

2011-06-16  David Bigagli  <david@openlava.net>

	* Makefiles changed all variables to run install from LAVA
	to OPENLAVA.

	* config/Makefile added a command to install COPYING and README files
	to the LAVA_ROOT/conf directory

2011-06-07  David Bigagli  <david@openlava.net>

	* lava startup script was rewritten, simplified and introduce
	the possibility to run daemons in single user mode.

	* Added a doc directory with original Platform pdf for lava
	administrator and user.

	* lava.spec Rober Stober added a new lava.spec to buidl rpm.

	* Make.misc removed the reference to the architecture x86_64
	and changed the make install to also install the lava sturtup script.


2011-05-02  David Bigagli  <david@openlava.net>

	* Makefile(s): Removed PREFIX variable from all make files and
	introduced Lava specific installation variables in Make.misc.

	* Makefile(s): Made static link for all Lava binaries.

	* Makefiles(s): Changed makefiles not to build shared libraries
	by default and modified the shared libraries to use standard .so
	suffix instead of .So.

2011-04-24  David Bigagli  <david@openlava.net>

	* ChangeLog: Created this file.

	* Make.def: Set the location of TCL header files in /usr/local/include
	instead of /usr/include. TCL_INCLUDE = /usr/local/include

	* Make.def: Added a static link for TCL libtcl8.4.a from
	/usr/local/lib. TCLLIB = /usr/local/lib/libtcl8.4.a

	* Make.def: Removed the -O2 optimization flag and the hash-style
	linker option.


lsf/lib/echkpnt.c
/* $Id: echkpnt.lib.c 397 2007-11-26 19:04:00Z mblack $
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include <unistd.h>
#include <strings.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <errno.h>
#include <time.h>

#include "echkpnt.lib.h"

extern char *putstr_ (const char *);
extern int initenv_ (struct config_param *, char *);

static FILE *m_pLogFile = NULL;
static char *m_pMessageHeader = NULL;
static char logMesgBuf[MAXLINELEN];

char *
getEchkpntMethodDir (char *pChkpnt_Cmd, const char *pMethodDir,
		     const char *pProgramName, const char *pMethodName)
{

  static char fname[] = "getEchkpntMethodDir()";
  char tmp_chkpnt_cmd[MAXPATHLEN];
  char *pConfigPath = NULL;
  char *pChkpntBuf = NULL;


  struct config_param aParamList[] = {
#define LSF_SERVERDIR    0
    {"LSF_SERVERDIR", NULL},
    {NULL, NULL}
  };


  if (pProgramName == NULL)
    {
      sprintf (logMesgBuf, "%s : parameter pProgramName is NULL\n", fname);
      fprintf (stderr, "%s", logMesgBuf);
      logMesg (logMesgBuf);
      return (NULL);
    }

  if (pChkpnt_Cmd == NULL)
    {
      pChkpntBuf = tmp_chkpnt_cmd;
    }
  else
    {
      pChkpntBuf = pChkpnt_Cmd;
    }


  pConfigPath = getenv ("LSF_ENVDIR");
  if (pConfigPath == NULL)
    {
      pConfigPath = "/etc";
    }
  if (initenv_ (aParamList, pConfigPath) < 0)
    {
      sprintf (logMesgBuf, "%s : initenv_() call fail\n", fname);
      logMesg (logMesgBuf);
    }

  if ((pMethodName != NULL)
      && (strcmp (pMethodName, ECHKPNT_DEFAULT_METHOD) != 0)
      && (pMethodDir != NULL))
    {
      strcpy (pChkpntBuf, pMethodDir);
    }
  else
    {
      if (aParamList[LSF_SERVERDIR].paramValue != NULL)
	{
	  strcpy (pChkpntBuf, aParamList[LSF_SERVERDIR].paramValue);
	}
      else
	{
	  sprintf (logMesgBuf, "%s : can't get the LSF_SERVERDIR value\n",
		   fname);
	  fprintf (stderr, "%s", logMesgBuf);
	  logMesg (logMesgBuf);
	  return (NULL);
	}
    }


  strcat (pChkpntBuf, "/");
  strcat (pChkpntBuf, pProgramName);
  strcat (pChkpntBuf, ".");


  if (pMethodName != NULL)
    {
      strcat (pChkpntBuf, pMethodName);
    }
  else
    {
      strcat (pChkpntBuf, ECHKPNT_DEFAULT_METHOD);
    }

  if (access (pChkpntBuf, X_OK) < 0)
    {
      sprintf (logMesgBuf, "%s : the %s is not executed \n%s\n",
	       fname, pChkpntBuf, errno ? strerror (errno) : "");
      fprintf (stderr, "%s", logMesgBuf);
      logMesg (logMesgBuf);
      return NULL;
    }

  if (pChkpnt_Cmd == NULL)
    {
      pChkpntBuf = putstr_ (pChkpntBuf);
      if (pChkpntBuf == NULL)
	{
	  sprintf (logMesgBuf, "%s : %s malloc dynamic space fail \n%s\n",
		   fname, "putstr_()", strerror (errno));
	  fprintf (stderr, "%s", logMesgBuf);
	  logMesg (logMesgBuf);
	}
    }
  if (aParamList[LSF_SERVERDIR].paramValue != NULL)
    {
      free (aParamList[LSF_SERVERDIR].paramValue);
    }
  return (pChkpntBuf);
}


int
getChkpntDirFile (char *pPathBuf, const char *pFileName)
{

  static char fname[] = "getChkpntDirFile()";
  char *pChkpntPath = NULL;
  int iIndex;

  if (pPathBuf == NULL)
    {
      return (-1);
    }
  strcpy (pPathBuf, "");

  pChkpntPath = getenv (ECHKPNT_CHKPNT_DIR);



  if (pChkpntPath != NULL)
    {
      sprintf (pPathBuf, "%s/", pChkpntPath);

#ifdef DEBUG
      sprintf (logMesgBuf, "%s : the LSB_CHKPNT_DIR is %s\n", fname,
	       pChkpntPath);
#endif

    }
  else
    {
      sprintf (logMesgBuf, "%s : LSB_CHKPNT_DIR is not defined\n", fname);
      logMesg (logMesgBuf);
      return (-1);
    }


  iIndex = strlen (pPathBuf);
  if (iIndex > 0)
    {
      while (pPathBuf[--iIndex] != '/');
      iIndex += 2;
      pPathBuf[iIndex] = '\0';
    }
  if (pFileName != NULL)
    {
      char tmpFileName[MAXFILENAMELEN];
      char *pFileNameChar = tmpFileName;
      strcpy (tmpFileName, pFileName);
      while (*(pFileNameChar) == '/')
	{
	  pFileNameChar++;
	}
      strcat (pPathBuf, pFileNameChar);
    }
  return (0);
}


int
redirectFd (const char *pFileName, int iFileNo)
{

  static char fname[] = "redirectFd()";
  int iFd;
  int ireValue = -1;

  if ((pFileName == NULL) || (iFileNo < 0))
    {
      return (-1);
    }


  iFd = open (pFileName, O_WRONLY | O_CREAT | O_APPEND, ECHKPNT_FILE_MODE);
  if (iFd == -1)
    {
      sprintf (logMesgBuf, "%s : open() can't open %s file\n%s\n",
	       fname, pFileName, errno ? strerror (errno) : "");
      logMesg (logMesgBuf);
      return (-1);
    }

  if (dup2 (iFd, iFileNo) != -1)
    {
      ireValue = 0;
    }
  else
    {
      sprintf (logMesgBuf, "%s : dup2() fail\n%s\n", fname, strerror (errno));
      logMesg (logMesgBuf);
      ireValue = -1;
    }
  if (iFd != iFileNo)
    {
      close (iFd);
    }
  return (ireValue);
}


int
initLog (char *pMesgHeader)
{

  char logFileName[MAXPATHLEN];

  if (m_pLogFile == NULL)
    {
      if (getChkpntDirFile (logFileName, ECHKPNT_LOG_FILENAME) != 0)
	{
	  return (-1);
	}
      else
	{
	  m_pLogFile = fopen (logFileName, "a");
	  if (m_pLogFile != NULL)
	    {

	      if (m_pMessageHeader != NULL)
		{
		  free (m_pMessageHeader);
		}
	      m_pMessageHeader = putstr_ (pMesgHeader);
	      fprintf (m_pLogFile,
		       "########### begin to checkpoint ############\n");
	      return (0);
	    }
	  return (-1);
	}
    }
  else
    {
      return (-1);
    }

}

void
setMesgHeader (char *pMesgHeader)
{
  if (m_pLogFile == NULL)
    {
      return;
    }
  if (m_pMessageHeader != NULL)
    {
      free (m_pMessageHeader);
    }
  m_pMessageHeader = putstr_ (pMesgHeader);
}


void
closeLog ()
{
  if (m_pLogFile != NULL)
    {
      fprintf (m_pLogFile, "########### %s end checkpoint ###########\n\n",
	       m_pMessageHeader != NULL ? m_pMessageHeader : "");
      fclose (m_pLogFile);
    }
  if (m_pMessageHeader != NULL)
    {
      free (m_pMessageHeader);
    }

}


void
logMesg (const char *pMessage)
{
  char timeString[MAX_TIME_STRING];

  time_t tTime;
  if ((m_pLogFile == NULL) || (pMessage == NULL))
    {
      return;
    }

  if (m_pMessageHeader != NULL)
    {
      strcpy (timeString, "");
      tTime = time (NULL);
      if (tTime != (time_t) - 1)
	{
	  int ii;
	  strcpy (timeString, ctime (&tTime));
	  ii = strlen (timeString);
	  if (timeString[ii - 1] == '\n')
	    {
	      timeString[ii - 1] = '\0';
	    }
	}
      fprintf (m_pLogFile, "%s : %s : %s", timeString, m_pMessageHeader,
	       pMessage);
      fflush (m_pLogFile);
    }

}
